`tidy.mvgam()`

swpease opened the pull request:
draft method

Some comments on the current status:
  - I haven't included the individual betas for random effects. I'm leaning towards that they should be included.
  - I think if using hierarchical residual correlation via `gr`, the Sigma matrix should be split out by group, but haven't implemented that yet.
  - Include the changepoints in piecewise trends?
  - I haven't yet considered an option to return the smooths significances, if that might be useful.
  - How to test?
    - The existing `mvgam_example[n]`s cover a good portion of possible outputs, but this method is supposed to handle every type of model. However, it seems excessive and minimally useful to cover that whole range with additional `mvgam_example`s.

Sorry this took to long -- I wanted to go over all the model types, and I've been managing eye strain so I can't just stare at text on my monitor for hours on end.

nicholasjclark reviewed (approved):
Thanks very much for this initial effort @swpease. I'm hugely appreciative, and overall I think this covers most of what users would like

swpease commented:
Alright, I think it's mostly there now. 

I still want to split out the different covariance matrixes for heirarchical correlation cases, but think it warrants having a new `mvgam_example` to check against. For `tidy`, the example model's output/sensibility is irrelevant, so long as it doesn't error, but maybe it would be good to have one that makes sense for other possible tests?

The best tests I could think of were just some snapshot tests. I opted for this unappealing JSON output because the regular output (a print) was giving technically-not-false errors due to varying output truncations. I can't say I've ever had seamless experiences with the testthat snapshotting capabilities.

swpease commented:
I implemented the Sigma splitting and added in a new "mvgam_example" (#6, for heirarchical correlated residuals models). I filtered the Sigmas by non-zero variance, which I believe does the job -- I had initially been going down the indexing tetris route.

swpease commented:
Hey @nicholasjclark just checking up on this. I have little idea of what actions trigger notifications to others. I meant to comment sooner but have been preoccupied with being crushed by the job application system as well as improving the snooker streaming experience.

swpease commented:
Oh, and any idea what's with the macos rstan failing check?

nicholasjclark reviewed (changes requested):
Thanks @swpease for these additions. Overall this looks good to me, but the only hesitancy I have is adding more example models in the internal `sysdata`. This is because each example takes up quite a lot of space and CRAN will reject packages if they are too large. Is it possible to do any of the testing without these internal examples, perhaps by just fitting some of the models within the tests themselves (like I do in `test-binomial.R` for example?)

swpease commented:
Oh, sure. I wasn't sure what made sense there. I had opted for making a new example model to avoid the added testing time. I'm a bit surprised that there doesn't seem to be a way to use `use_data` for test-specific data that doesn't go into the package.

nicholasjclark commented:
Hi @swpease, these failures aren't a major issue, but do you need to use `backend = 'rstan'` in the example model for a specific reason? I suggest removing that so that we don't have to install all the `rstan` dependencies in the main testing workflow

swpease commented:
I just picked rstan because I couldn't figure out how to set a seed in cmdstanr. 

In any case, is the "test/" directory part of CRAN package size considerations? I was hoping to be able to store a "test_data" file there, which would make things much cleaner.

nicholasjclark commented:
I believe the `test` directory is considered, unfortunately. But it isn't a problem to run the model each time we test, in general this is useful to do anyway and I only ever do it through Github actions. But if you need a seed to ensure numerical stability then we can always just install `rstan` dependencies on the main CMDcheck workflow (https://github.com/nicholasjclark/mvgam/blob/master/.github/workflows/R-CMD-check.yaml). I believe that all we need are               `BH` and `RcppEigen`, as `rstan` is already one of our dependencies. The checks should then work just fine

swpease commented:
Darn. Well, what if I just ignored the numerical portion of the output I'm testing? The non-numerical stuff (param name/type and column names) should I believe cover the responsibilities of `tidy`.

nicholasjclark commented:
Great to see all the tests passing now @swpease. Are you planning in more edits or should I go ahead and merge?

swpease commented:
Yeah, go ahead a merge.
